##  Final Pre-Processing Dataset

In [1]:
# Ensure working directory is set to project root
import os
if os.path.basename(os.getcwd()) == "notebooks":
    os.chdir(".. ".strip())

# Import pandas for DataFrame handling
import pandas as pd

# Load engineered dataset
engineered_csv_path = "data/processed/utkface_engineered.csv"
df = pd.read_csv(engineered_csv_path)

print(f"Successfully loaded '{engineered_csv_path}' ({len(df)} rows, {len(df.columns)} columns).")

Successfully loaded 'data/processed/utkface_engineered.csv' (2998 rows, 17 columns).


In [2]:
# Exclude non-predictive identifier columns ('image_name', 'filepath')
# Reorder columns so target variable 'age' appears first for readability
final_columns = [
    'age', 'gender_0', 'gender_1', 'race_0', 'race_1', 'race_2', 'race_3', 'race_4',
    'eye_distance', 'mouth_width', 'face_width', 'face_height',
    'face_aspect_ratio', 'eye_to_face_ratio', 'mouth_to_face_ratio'
]

final_df = df[final_columns].copy()
print(f"Created final_df without identifier columns. Shape: {final_df.shape}")

Created final_df without identifier columns. Shape: (2998, 15)


In [3]:
print("=== FINAL DATAFRAME HEAD (FIRST 10 ROWS) ===")
print(final_df.head(10))

print("\n=== FINAL DATAFRAME SHAPE ===")
print(f"Shape (rows, columns): {final_df.shape}")

print("\n=== FINAL DATAFRAME INFO & MEMORY FOOTPRINT ===")
final_df.info()

print("\n=== FINAL MISSING VALUE CHECK (ISNULL().SUM()) ===")
print(final_df.isnull().sum())

=== FINAL DATAFRAME HEAD (FIRST 10 ROWS) ===
   age  gender_0  gender_1  race_0  race_1  race_2  race_3  race_4  \
0   21         1         0       0       0       1       0       0   
1   39         1         0       0       0       0       0       1   
2   36         1         0       0       0       0       0       1   
3   26         1         0       0       0       0       1       0   
4   21         0         1       0       1       0       0       0   
5   61         0         1       0       0       0       1       0   
6   54         1         0       0       1       0       0       0   
7   25         1         0       0       0       0       1       0   
8   35         1         0       0       0       0       1       0   
9   39         0         1       1       0       0       0       0   

   eye_distance  mouth_width  face_width  face_height  face_aspect_ratio  \
0     77.408943    66.862511  160.524368   193.195435           1.203527   
1     81.073242    68.352386  17

**Screenshot Instructions:** Take a screenshot of the `final_df.head(10)` output above (showing column names, sample rows, and confirm it visually represents the final cleaned/transformed state) for CE1 submission.

In [4]:
# Save final_df to data/processed/utkface_final_preprocessed.csv
final_csv_path = "data/processed/utkface_final_preprocessed.csv"
final_df.to_csv(final_csv_path, index=False)

print(f"Final preprocessed dataset saved successfully to '{final_csv_path}' ({len(final_df)} rows, {len(final_df.columns)} columns).")

Final preprocessed dataset saved successfully to 'data/processed/utkface_final_preprocessed.csv' (2998 rows, 15 columns).


## Final CE1 Requirement Summary

| Requirement | Implementation | Status |
|:---|:---|:---|
| **Data Collection** | UTKFace dataset (23,708 images), metadata parsed from filenames | Completed |
| **Data Cleaning** | Duplicate/invalid-value/broken-filepath checks (0 issues found), 3 malformed filenames excluded | Completed |
| **Missing Values** | `isnull()` check before/after (0 missing values found, no treatment needed) | Completed |
| **Categorical Encoding** | One-Hot Encoding applied to gender and race | Completed |
| **Data Integration** | Not applicable - no separate landmark file existed to merge; landmarks were extracted directly from images instead | Not Applicable (documented) |
| **Feature Extraction** | Facial landmarks extracted via OpenCV YuNet on a documented 3000-image random sample (2998 successful) | Completed |
| **Feature Engineering** | 3 ratio features derived (`face_aspect_ratio`, `eye_to_face_ratio`, `mouth_to_face_ratio`) | Completed |
| **Feature Selection** | X (14 features) and y (age) explicitly defined | Completed |
| **EDA & Visualization** | 5 required graphs created with real observations | Completed |
| **Final Preprocessed Dataset** | Cleaned, encoded, feature-engineered final DataFrame (2998 rows x 15 columns) | Completed |